internal/database/badger_gc.go

```diff
@@ -0,0 +1,99 @@
+/*
+Copyright 2025 The Flux authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package database
+
+import (
+	"errors"
+	"time"
+
+	"github.com/dgraph-io/badger/v3"
+	"github.com/go-logr/logr"
+)
+
+type BadgerGarbageCollector struct {
+	// settings
+	DiscardRatio float64
+	Interval     time.Duration
+	// external deps
+	db  *badger.DB
+	log *logr.Logger
+	// flow control
+	timer *time.Timer
+	stop  chan struct{}
+}
+
+// NewBadgerGarbageCollector creates and returns a new
+func NewBadgerGarbageCollector(db *badger.DB, interval time.Duration, log *logr.Logger) *BadgerGarbageCollector {
+	return &BadgerGarbageCollector{
+		DiscardRatio: 0.5, // must be a float between 0.0 and 1.0, inclusive
+		Interval:     interval,
+
+		db:  db,
+		log: log,
+
+		timer: time.NewTimer(interval),
+		stop:  make(chan struct{}),
+	}
+}
+
+// Run repeatedly runs the BadgerDB garbage collector with a delay inbetween
+// runs.
+//
+// This is a blocking operation, so it should be run as a separate goroutine.
+// To stop the garbage collector, call Stop().
+func (gc *BadgerGarbageCollector) Run() {
+	gc.log.Info("Starting Badger GC")
+	for {
+		select {
+		case <-gc.timer.C:
+			gc.discardValueLogFiles()
+			gc.timer.Reset(gc.Interval)
+		case <-gc.stop:
+			gc.timer.Stop()
+			gc.log.Info("Stopped Badger GC")
+			gc.stop <- struct{}{}
+			return
+		}
+	}
+}
+
+// Stop blocks until the garbage collector has been stopped.
+//
+// To avoid GC Errors, call Stop() before closing the database.
+func (gc *BadgerGarbageCollector) Stop() {
+	gc.log.Info("Sending stop to Badger GC")
+	gc.stop <- struct{}{}
+	<-gc.stop
+}
+
+// upper bound for loop
+const maxDiscards = 1000
```

Comment:
I don't know if this is a reasonable value of discards, but it felt like the controller could sample, delete, and rewrite 1000 value log files relatively quickly in its own goroutine without impacting the controller-runtime too much.